src/coins.h

```diff
@@ -103,6 +106,30 @@ class Coin
  */
 struct CCoinsCacheEntry
 {
+private:
+    /**
+     * These are used to create a doubly linked list of flagged entries.
+     * They are set in AddFlags and unset in ClearFlags.
+     * A flagged entry is any entry that is either DIRTY, FRESH, or both.
+     *
+     * DIRTY entries are tracked so that only modified entries can be passed to
+     * the parent cache to do a batch write. This is a performance optimization
+     * compared to giving all entries in the cache to the parent and having the
+     * parent scan for only modified entries.
+     *
+     * FRESH entries are tracked because a FRESH-but-not-DIRTY entry can only be
+     * a spent coin that exists in the parent cache. When performing a batch
```

Comment:
> With that, this PR could add only DIRTY coins and ignore the FRESH flag, which would be simpler.

I'm not sure how that would be simpler. The two flags are coupled and so adding one and not the other would be a more complex diff. Perhaps I don't understand what you are suggesting?

I could just replace the single FRESH-but-not-DIRTY check with an assertion and remove this comment about tracking FRESH coins?

Reply:
Yes, I thought some more and don't find a great way to simplify the logic. I think it would be enough to remove / amend the comment, maybe mentioning instead that FRESH-but-not-DIRTY coins are not expected to occur but would be handled well if they did.

> replace the single FRESH-but-not-DIRTY check 

Not sure if that's the "check" you mean, but replacing [this](https://github.com/bitcoin/bitcoin/blob/b27afb7fb774809c9c075d56e44657e0b607a00c/src/coins.cpp#L51-L55) with an assertion was essentially what #18746 did in its latest version. Since that PR was mildly controversial (I think mostly because it originally had the spin of "These coins may be unnecessarily added today; don't add them anymore" instead of "We already don't add them. Remove unreachable code") I think the question of adding an assert should be separated from this PR.

Reply:
Updated the comment.

Reply:
Revived in #30673